src/d3d9/d3d9_device.cpp

```diff
@@ -1953,12 +1955,47 @@ namespace dxvk {
   }
 
   HRESULT STDMETHODCALLTYPE Direct3DDevice9Ex::SetStreamSourceFreq(UINT StreamNumber, UINT Setting) {
-    Logger::warn("Direct3DDevice9Ex::SetStreamSourceFreq: Stub");
+    auto lock = LockDevice();
+
+    if (StreamNumber >= caps::MaxStreams
+      || Setting == 0
+      || (Setting & D3DSTREAMSOURCE_INDEXEDDATA) == D3DSTREAMSOURCE_INDEXEDDATA && (Setting & D3DSTREAMSOURCE_INSTANCEDATA == D3DSTREAMSOURCE_INSTANCEDATA)
+      || (Setting & D3DSTREAMSOURCE_INDEXEDDATA) == D3DSTREAMSOURCE_INDEXEDDATA && StreamNumber != 0 //IndexData must always be stream 0
+      || (Setting & D3DSTREAMSOURCE_INSTANCEDATA) == D3DSTREAMSOURCE_INSTANCEDATA && StreamNumber == 0) //InstanceData must not be stream 0
+      return D3DERR_INVALIDCALL;
+
+    if (unlikely(ShouldRecord()))
+      return m_recorder->SetStreamSourceFreq(
+        StreamNumber,
+        Setting);
+
+    if ((Setting & D3DSTREAMSOURCE_INDEXEDDATA) == D3DSTREAMSOURCE_INDEXEDDATA)
```

Comment:
Your tracking of the state here is incorrect and won't work.

Make the state that we store the values the user give so we can return them correctly in GetStreamSourceFreq, then resolve what we have to in the main device, outside of what we consider the state to be.